### PR TITLE
Add documentation for WSL

### PR DESCRIPTION
**Pull request by pieandcakes**
@bobbrow @caslan @rajkumar42 @ronglums Please review

### CODE REVIEW

**Review by rajkumar42 — APPROVED** [review]

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+**NOTE: Creator's Update is required due to bugfixes within the subsystem that we rely on to provide debugging. Debugging using a previous version of WSL is unsupported and likely will not work.**
+
+## Prerequisites
+* [Windows 10 Creator's Update with Windows Subsystem for Linux and Bash.](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide)
+* Install g++/gcc and gdb to allow compiling and debugging.
+* [Visual Studio Code](https://code.visualstudio.com) + Microsoft C/C++ extension for VSCode.
+
+## How-To
+To debug, commands will be routed from Windows through `bash.exe` to setup debugging. Because our extension runs as a 32-bit process, it will need to use the `C:\Windows\SysNative` folder to access the executable that is normally in `C:\Windows\System32`. We will be using the `"pipeTransport"` ability within the extension to do debugging and `"sourceFileMap"` to map the source from the subsystem's paths back to Windows path.
```
Comment (WardenGnaw) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: Which executable?

Comment (pieandcakes) [reply Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: bash.exe . I'll add that in.

**Review by WardenGnaw — APPROVED** [review]

**Comment by timsneath** [comment]
Do we want to include anything about configuration of `c_cpp_properties.json` so that Linux header files are found instead of Windows ones? 

I included the following in mine:
```
"includePath": [
                "${localappdata}/lxss/rootfs/usr/include",
                "${localappdata}/lxss/rootfs/usr/local/include"            
            ],
```

@bitcrazed pointed out that this could _theoretically_ be risky. One alternative could be to tell users to copy header files from /usr/include to a location outside of the lxss folder and use that for the includePath, but the problem there is that header updates won't get automatically pulled through. I've stuck with the original solution for now though.

**Comment by pieandcakes** [comment]
@timsneath Thanks for the info. I work on the debugging experience but hopefully @bobbrow @ronglums @sean-mcmanus can chime in. Are there plans for additional instructions/configurations from a Language Server standpoint that you will be providing to help this experience?

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# Windows 10's Windows Subsystem for Linux
+With the release of Windows 10 Creator's Update, you will now be able to use Visual Studio Code and the Microsoft C/C++ extension to debug your `Windows Subsystem for Linux (WSL)` Bash on Ubuntu projects.
```
Comment (timsneath) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: I think the official name is "Windows 10 Creators Update" (no apostrophe) - see https://blogs.windows.com/buildingapps/2017/04/05/windows-10-creators-update-creators-update-sdk-released/ for an example.

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+Code can be written on Windows itself using VSCode and debugged through `bash.exe` to the Bash on Windows layer. 
+
+**NOTE: Creator's Update is required due to bugfixes within the subsystem that we rely on to provide debugging. Debugging using a previous version of WSL is unsupported and likely will not work.**
+
+## Prerequisites
+* [Windows 10 Creator's Update with Windows Subsystem for Linux and Bash.](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide)
+* Install g++/gcc and gdb to allow compiling and debugging.
```
Comment (timsneath) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: In WSL, right? Not in Windows? Probably want to make that really clear. Perhaps worth calling out `sudo apt get build-essential` as the right command to get stuff set up, unless you think the audience will already know all this.

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+* [Visual Studio Code](https://code.visualstudio.com) + Microsoft C/C++ extension for VSCode.
+
+## How-To
+To debug, commands will be routed from Windows through `bash.exe` to setup debugging. Because our extension runs as a 32-bit process, it will need to use the `C:\Windows\SysNative` folder to access the `bash.exe` executable that is normally in `C:\Windows\System32`. We will be using the `"pipeTransport"` ability within the extension to do debugging and `"sourceFileMap"` to map the source from the subsystem's paths back to Windows path. 
+
+**NOTE: Applications will need to be compiled in the `Windows Subsystem for Linux (WSL)` prior to debugging.**
+
```
Comment (timsneath) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: Probably worth a link back to https://code.visualstudio.com/Docs/editor/debugging so users who aren't familiar with launch.json and debugging in Visual Studio Code in general have a point of reference...? This page also explains how to create a launch.json and the various elements contained within it.

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                "pipeArgs": ["-c"],
+                "debuggerPath": "/usr/bin/gdb"
+            },
+            "sourceFileMap": {
+                "/mnt/z": "z:\\"
+            }
+        }
+```
+
+### Example `launch.json` for Attaching to an Existing Process
```
Comment (timsneath) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md:55]: Maybe I'm dumb, but I don't understand when I'd use the latter? This is like Debug / Attach to Process in 'full' VS, right?

**Review by timsneath — COMMENTED** [review]
Few comments enclosed...

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+**NOTE: Creator's Update is required due to bugfixes within the subsystem that we rely on to provide debugging. Debugging using a previous version of WSL is unsupported and likely will not work.**
+
+## Prerequisites
+* [Windows 10 Creator's Update with Windows Subsystem for Linux and Bash.](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide)
+* Install g++/gcc and gdb to allow compiling and debugging.
+* [Visual Studio Code](https://code.visualstudio.com) + Microsoft C/C++ extension for VSCode.
+
+## How-To
+To debug, commands will be routed from Windows through `bash.exe` to setup debugging. Because our extension runs as a 32-bit process, it will need to use the `C:\Windows\SysNative` folder to access the `bash.exe` executable that is normally in `C:\Windows\System32`. We will be using the `"pipeTransport"` ability within the extension to do debugging and `"sourceFileMap"` to map the source from the subsystem's paths back to Windows path.
```
Comment (bobbrow) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: `setup` is used as a verb a few times in this document, but when spelled as one word it is a noun.  Replace with `set up` (two words)

Comment (bobbrow) [inline Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: `setup` is used as a verb a few times in this document, but when written as a single word, it is actually a noun.  Replaced with `set up` (two words)

**Comment by bitcrazed** [comment]
@robotdad Are you able to shed some light on how VC++ for Linux handles headers & Intellisense?

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md (hunk 4 above)
Comment (pieandcakes) [reply Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: @timsneath This isn't a tutorial but an addition of documentation within our repo for debugging. As such, I'm assuming that the user knows launch.json

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md (hunk 5 above)
Comment (pieandcakes) [reply Documentation/Debugger/gdb/Windows Subsystem for Linux.md:55]: If you have a process already running in Bash on Windows and want to debug it, this is how you do it.

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md (hunk 2 above)
Comment (pieandcakes) [reply Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: Fixed

**Comment by robotdad** [comment]
For VC Linux support we carry basic IntelliSense for libstdc++, it isn't tailored for any specific Linux system. To get IntelliSense specific to your Linux system you need to copy your include files (or map them to a Windows drive etc.) and give that local path to VS to parse them. That's described [here](https://blogs.msdn.microsoft.com/vcblog/2016/03/30/visual-c-for-linux-development/#includes)

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md (hunk 3 above)
Comment (pieandcakes) [reply Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: Added `WSL` to the sentence. The command you provided isn't correct but i'll add something in regards to using the package manager to install tools.

File: Documentation/Debugger/gdb/Windows Subsystem for Linux.md (hunk 6 above)
Comment (pieandcakes) [reply Documentation/Debugger/gdb/Windows Subsystem for Linux.md]: Fixed

**Comment by bitcrazed** [comment]
🍺 to Marc for NOT instructing users to copy files from %localappdata%\lxss

Using [p]scp to copy files from Linux is a MUCH better bet.

**Pull request merged**